Update GraphTask::owner_ in a single thread for DistEngine.

pritamdamania87 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#58625 Update GraphTask::owner_ in a single thread for DistEngine.**

Several TSAN tests were failing for distributed since `owner_` was not
atomic and was being accessed by several threads. As an example:
https://github.com/pytorch/pytorch/blob/master/torch/csrc/distributed/autograd/engine/dist_engine.cpp#L333.

To fix this, I've set the owner_ only once when the graphTask is created.

Differential Revision: [D28496878](https://our.internmc.facebook.com/intern/diff/D28496878/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D28496878/)!

pritamdamania committed:
Update GraphTask::owner_ in a single thread for DistEngine.

Several TSAN tests were failing for distributed since `owner_` was not
atomic and was being accessed by several threads. As an example:
https://github.com/pytorch/pytorch/blob/master/torch/csrc/distributed/autograd/engine/dist_engine.cpp#L333.

To fix this, I've set the owner_ only once when the graphTask is created.

Differential Revision: [D28496878](https://our.internmc.facebook.com/intern/diff/D28496878/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D28496878/)!

[ghstack-poisoned]

albanD reviewed (approved):
LGTM

Thanks for the udpated fix!

pritamdamania committed:
Update on "Update GraphTask::owner_ in a single thread for DistEngine."


Several TSAN tests were failing for distributed since `owner_` was not
atomic and was being accessed by several threads. As an example:
https://github.com/pytorch/pytorch/blob/master/torch/csrc/distributed/autograd/engine/dist_engine.cpp#L333.

To fix this, I've set the owner_ only once when the graphTask is created.

Differential Revision: [D28496878](https://our.internmc.facebook.com/intern/diff/D28496878/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D28496878/)!

[ghstack-poisoned]
